server/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1741,4 +1605,98 @@ public final ResponseHandlers getResponseHandlers() {
     public final RequestHandlerRegistry<? extends TransportRequest> getRequestHandler(String action) {
         return requestHandlers.get(action);
     }
+
+    private final class ChannelsConnectedListener implements ActionListener<Void> {
+
+        private static final int FAILED = -1;
+
+        private final DiscoveryNode node;
+        private final ConnectionProfile connectionProfile;
+        private final List<TcpChannel> channels;
+        private final ActionListener<NodeChannels> listener;
+        private final AtomicInteger pendingConnections;
```

Comment:
I'd use `CountDown.java` here instead for the `pendingConnections` and an extra boolean if we failed or not. That is more straight forward and the impls will be simpler.